src/test/kotlin/org/rust/ide/intentions/ConvertMethodCallToUFCSTest.kt

```diff
@@ -0,0 +1,360 @@
+        struct Foo;
+        impl Foo {
+            fn foo(&self) {}
+        }
+        fn foo(foo: Foo) {
+            foo./*caret*/foo();
+        }
+    """, """
+        struct Foo;
+        impl Foo {
+            fn foo(&self) {}
+        }
+        fn foo(foo: Foo) {
+            Foo::foo(&foo);
+        }
+    """)
+
+    fun `test self ref receiver ref`() = doAvailableTest("""
+        struct Foo;
+        impl Foo {
+            fn foo(&self) {}
+        }
+        fn foo(foo: &Foo) {
+            foo./*caret*/foo();
+        }
+    """, """
+        struct Foo;
+        impl Foo {
+            fn foo(&self) {}
+        }
+        fn foo(foo: &Foo) {
+            Foo::foo(foo);
+        }
+    """)
+
+    fun `test self mut ref receiver owned`() = doAvailableTest("""
+        struct Foo;
+        impl Foo {
+            fn foo(&mut self) {}
+        }
+        fn foo(mut foo: Foo) {
+            foo./*caret*/foo();
+        }
+    """, """
+        struct Foo;
+        impl Foo {
+            fn foo(&mut self) {}
+        }
+        fn foo(mut foo: Foo) {
+            Foo::foo(&mut foo);
+        }
+    """)
+
+    fun `test self mut ref receiver mut ref`() = doAvailableTest("""
+        struct Foo;
+        impl Foo {
+            fn foo(&mut self) {}
+        }
+        fn foo(foo: &mut Foo) {
+            foo./*caret*/foo();
+        }
+    """, """
+        struct Foo;
+        impl Foo {
+            fn foo(&mut self) {}
+        }
+        fn foo(foo: &mut Foo) {
+            Foo::foo(foo);
+        }
+    """)
+
+    fun `test nested attribute method call`() = doAvailableTest("""
+        struct Bar;
+        struct Foo {
+            bar: Bar
+        }
+        impl Bar {
+            fn bar(&self) {}
+        }
+        fn foo(foo: &Foo) {
+            foo.bar./*caret*/bar();
+        }
+    """, """
+        struct Bar;
+        struct Foo {
+            bar: Bar
+        }
+        impl Bar {
+            fn bar(&self) {}
+        }
+        fn foo(foo: &Foo) {
+            Bar::bar(&foo.bar);
+        }
+    """)
+
+    fun `test trait method call on struct`() = doAvailableTest("""
+        struct Foo;
+        trait Trait {
+            fn foo(&self);
+        }
+        impl Trait for Foo {
+            fn foo(&self) {}
+        }
+
+        fn foo(foo: &Foo) {
+            foo./*caret*/foo();
+        }
+    """, """
+        struct Foo;
+        trait Trait {
+            fn foo(&self);
+        }
+        impl Trait for Foo {
+            fn foo(&self) {}
+        }
+
+        fn foo(foo: &Foo) {
+            Foo::foo(foo);
+        }
+    """)
+
+    fun `test trait method call on generic trait bound`() = doAvailableTest("""
+        trait Trait {
+            fn foo(&self);
+        }
+
+        fn foo<T: Trait>(foo: &T) {
+            foo./*caret*/foo();
+        }
+    """, """
+        trait Trait {
+            fn foo(&self);
+        }
+
+        fn foo<T: Trait>(foo: &T) {
+            Trait::foo(foo);
```

Comment:
Why not `T::foo`?

Reply:
I wanted to do that, but didn't know how to get the type parameter elegantly. I switched to rendering the `selfTy` of the variant in most cases.